File: problems/1401-1500/1474_prune-a-linked-list-in-keep-drop-cycles/solution_values.go
// Sweep 1 records the values that survive each keep-m, drop-n cycle; a
// keep run cut short by the tail simply ends the walk, and a drop run
// steps past the nodes it loses. Sweep 2 rebuilds a fresh list from the
// surviving values.
func pruneNodes(head *ListNode, m int, n int) *ListNode {
	kept := []int{}
	node := head
	for node != nil {
		for i := 0; i < m && node != nil; i++ {
			kept = append(kept, node.Val)
			node = node.Next
		}
		for i := 0; i < n && node != nil; i++ {
			node = node.Next
		}
	}
	dummy := &ListNode{Val: 0}
	tail := dummy
	for _, value := range kept {
		tail.Next = &ListNode{Val: value}
		tail = tail.Next
	}
	return dummy.Next
}

File: problems-adapt/1401-1500/1474_prune-a-linked-list-in-keep-drop-cycles/solution_values.rs
impl Solution {
    pub fn prune_nodes(head: Option<Box<ListNode>>, m: i32, n: i32) -> Option<Box<ListNode>> {
        // Sweep 1: record the values that survive each keep-m, drop-n cycle.
        // The walk peeks by shared reference and never rewrites a link, so
        // the input list is only read.
        let mut kept: Vec<i32> = Vec::new();
        let mut node = head.as_deref();
        while node.is_some() {
            for _ in 0..m {
                match node {
                    Some(run) => {
                        kept.push(run.val);
                        node = run.next.as_deref();
                    }
                    None => break,
                }
            }
            for _ in 0..n {
                match node {
                    Some(run) => node = run.next.as_deref(),
                    None => break,
                }
            }
        }
        // Sweep 2: rebuild a fresh list threaded from the surviving values.
        let mut dummy = Box::new(ListNode { val: 0, next: None });
        let mut tail = dummy.as_mut();
        for value in kept {
            tail.next = Some(Box::new(ListNode { val: value, next: None }));
            tail = tail.next.as_mut().unwrap();
        }
        dummy.next
    }
}
